### PR TITLE
Fix #1685. Correct handling of SqlProviderFlags.IsParameterOrderDependent.

### LINKED ISSUE #1685: Sql.Expression doesn't respect providers with positional parameters
We have several tests for SAP HANA with same failure as in #1528, caused by different number and order of parameters in SQL and command parameters collection.

E.g. `IndexOf3` test produce following SQL:
```sql
SELECT
	"p"."FirstName",
	"p"."PersonID",
	"p"."LastName",
	"p"."MiddleName",
	"p"."Gender"
FROM
	"Person" "p"
WHERE
	CASE
		WHEN :Length0 = 0 AND Length("p"."LastName") > :n3
			THEN :n3
		ELSE Locate(Substring(Left("p"."LastName", :n2),:n4 + 1),:s) + :n4 - 1
	END = 1 AND
	"p"."PersonID" = 2
```
with following parameters: 

Actual: `Length0, n3, s, n2, n4`

Expected: `Length0, n3, n3, n2, n4, s, n4`

### PR DESCRIPTION
**Pull request by sdanyliv**
Fix #1685

### CODE REVIEW

**Comment by MaceWindu** [comment]
Postponse. Blocked by #1512

File: Source/LinqToDB/SqlQuery/QueryVisitor.cs [hunk 1]
```diff
@@ -2305,5 +2306,975 @@ List<T> Convert<T>(IList<T> list1, Clone<T> clone)
...
+						var target     = (SqlTableSource)ConvertImmutableInternal(merge.Target);
+						var source     = (SqlMergeSourceTable)ConvertImmutableInternal(merge.Source);
+						var on         = (SqlSearchCondition)ConvertImmutableInternal(merge.On);
+						var operations = ConvertImmutableSafe(merge.Operations);
+
+						if (target     != null && !ReferenceEquals(merge.Target, target) ||
+							source     != null && !ReferenceEquals(merge.Source, source) ||
+							on         != null && !ReferenceEquals(merge.On, on) ||
+							operations != null && !ReferenceEquals(merge.Operations, operations))
+							newElement = new SqlMergeStatement(
```
Comment (sdanyliv) [inline Source/LinqToDB/SqlQuery/QueryVisitor.cs]: For good readability, put brackets here.

File: Source/LinqToDB/SqlQuery/QueryVisitor.cs [hunk 2]
```diff
@@ -2305,5 +2306,975 @@ List<T> Convert<T>(IList<T> list1, Clone<T> clone)
...
+						var source = (SqlMergeSourceTable)element;
+
+						var enumerableSource = (SqlValuesTable)ConvertImmutableInternal(source.SourceEnumerable);
+						var querySource      = (SelectQuery)ConvertImmutableInternal(source.SourceQuery);
+						var fields           = ConvertImmutableSafe(source.SourceFields);
+
+						if (enumerableSource != null && !ReferenceEquals(source.SourceEnumerable, enumerableSource) ||
+							querySource      != null && !ReferenceEquals(source.SourceQuery, querySource)           ||
+							fields           != null && !ReferenceEquals(source.SourceFields, fields))
+							newElement = new SqlMergeSourceTable(
```
Comment (sdanyliv) [inline Source/LinqToDB/SqlQuery/QueryVisitor.cs]: For good readability, put brackets here.

File: Source/LinqToDB/SqlQuery/QueryVisitor.cs [hunk 3]
```diff
@@ -2305,5 +2306,975 @@ List<T> Convert<T>(IList<T> list1, Clone<T> clone)
...
+								{
+									var field = fields2[i];
+
+									fields2[i] = new SqlField(field);
+
+									_visitedElements[field] = fields2[i];
+								}
+							}
+
+						newElement = new SqlValuesTable(fields2, rowsConverted ? covertedRows : table.Rows);
```
Comment (sdanyliv) [inline Source/LinqToDB/SqlQuery/QueryVisitor.cs]: Here always generated new value.

File: Source/LinqToDB/SqlQuery/QueryVisitor.cs [hunk 4]
```diff
@@ -2305,5 +2306,975 @@ List<T> Convert<T>(IList<T> list1, Clone<T> clone)
...
+						var operation = (SqlMergeOperationClause)element;
+
+						var where       = (SqlSearchCondition)ConvertImmutableInternal(operation.Where);
+						var whereDelete = (SqlSearchCondition)ConvertImmutableInternal(operation.WhereDelete);
+						var items       = ConvertImmutableSafe(operation.Items);
+
+						if (where       != null && !ReferenceEquals(operation.Where, where)             ||
+							whereDelete != null && !ReferenceEquals(operation.WhereDelete, whereDelete) ||
+							items       != null && !ReferenceEquals(operation.Items, items))
+							newElement = new SqlMergeOperationClause(
```
Comment (sdanyliv) [inline Source/LinqToDB/SqlQuery/QueryVisitor.cs]: For good readability, put brackets here.

File: Source/LinqToDB/SqlQuery/QueryVisitor.cs [hunk 5]
```diff
@@ -2305,5 +2306,975 @@ List<T> Convert<T>(IList<T> list1, Clone<T> clone)
...
+				case QueryElementType.TruncateTableStatement:
+					{
+						var truncate = (SqlTruncateTableStatement)element;
+
+						if (truncate.Table != null)
+						{
+							var table = (SqlTable)ConvertImmutableInternal(truncate.Table);
+
+							if (table != null && !ReferenceEquals(truncate.Table, table))
+								newElement = new SqlTruncateTableStatement()
```
Comment (sdanyliv) [inline Source/LinqToDB/SqlQuery/QueryVisitor.cs]: For good readability, put brackets here.

**Pull request merged**